### DISCUSSION TITLE
P2002 Error Code and Bugs

### DISCUSSION THREAD

**Discussion by periakteon** [prisma/prisma#19227] [Q&A]
I have a register API that I wrote using Next.js, Prisma, Zod and TypeScript.

I can manually register as many users as I want. There is no problem. As you can imagine, it can be very difficult to test manually. So I wrote a Node.js script for this, but I'm running into a problem.

First of all, let me write the operations I do in order:

i) I create random users using Faker.js and save them in \`users.json\` file.

```js
import fs from "fs";
import { faker } from "@faker-js/faker";

const generateUser = () => {

  // Create an array to store the users
  const users = [];

  // Create a set to store the used emails
  const usedEmails = new Set();
  // Generate 100 random users
  for (let i = 0; i < 100; i++) {
    // Generate a random username
    let username = faker.name.firstName();

    // Check if the username has been used before
    while (usedEmails.has(username)) {
      console.log(
        `Warning: ${email} has already been used, new e-mail is being created.`
      );
      username = faker.name.firstName();
    }

    // Generate a random email
    const email = faker.internet.email();

    // Check if the email has been used before
    if (usedEmails.has(email)) {
      console.log(
        `Warning: ${email} has already been used, new e-mail is being created.`
      );
      email = faker.internet.email();
    }

    // Generate a random password
    const password = faker.internet.password();

    // Check if the username is at least 5 characters long
    while (username.length < 5) {
      username = faker.name.firstName();
    }

    // Create a user object
    const user = {
      username,
      email,
      password,
    };

    // Log the user creation
    console.log(`User has been created: ${i}`);

    // Add the user to the users array
    users.push(user);
  }
  // Log the duplicated emails
  console.log("Duplicated e-mails:", usedEmails);

  // Convert the users array to JSON
  const jsonUsers = JSON.stringify(users);

  // Write the users to the users.json file
  fs.writeFile("users.json", jsonUsers, (err) => {
    if (err) {
      console.error(err);
      return;
    }
    console.log("Users has been saved to users.json");
  });
};

// Generate the users
generateUser();
```

ii) I send each user in the \`users.json\` file to the register API via a script with the POST method.

```js
import users from "./users.json" assert { type: "json" };
import axios from "axios";

export const registerUsers = async (users) => {
  const failedEmails = [];

  for (let i = 0; i < users.length; i++) {
    const user = users[i];

    const data = {
      username: user.username,
      email: user.email.trim(),
      password: user.password,
    };

    try {
      const response = await axios.post(
        "http://localhost:3000/api/auth/register",
        data
      );
      console.log(`User ${i} registered:`, response.data);
    } catch (error) {
      console.error(
        `There is an error while trying to register ${i} (${user.email})`,
        error.response.data
      );
      if (error.response.data.errors.includes("This e-mail is already taken.")) {
        failedEmails.push(user.email);
      }
    }

    await new Promise((resolve) => setTimeout(resolve, 1));
  }

  console.log("All users have been registered!");
  console.log(`Failed e-mails (${failedEmails.length}): ${failedEmails.join(", ")}`);
};

registerUsers(users);
```

This is exactly where the problem starts. There are no duplicated emails in the \`users.json\` file, but when I run the register script, at least a few of the users return errors from the server every time.

My register API code:

```ts
import type { NextApiRequest, NextApiResponse } from "next";
import { Prisma, PrismaClient } from "@prisma/client";
import crypto from "crypto";
import { z } from "zod";
import { hashPassword } from "@/utils/utils";
import { registerResponse, registerRequestSchema } from "@/utils/zodSchemas";

const prisma = new PrismaClient();
type ResponseType = z.infer<typeof registerResponse>;

export default async function handleRegister(
  req: NextApiRequest,
  res: NextApiResponse<ResponseType>,
) {
  if (req.method !== "POST") {
    return res.status(405).json({ success: false, errors: ["Method not allowed"] });
  }

  const parsed = await registerRequestSchema.safeParseAsync(req.body);

  if (!parsed.success) {
    const errorMap = parsed.error.flatten().fieldErrors;
    const errorMessages = Object.values(errorMap).flatMap(
      (errors) => errors ?? [],
    );

    return res.status(400).json({ success: false, errors: errorMessages });
  }

  const { username, email, password } = parsed.data;

  const { salt, hash } = hashPassword(
    password,
    crypto.randomBytes(16).toString("hex"),
  );

  try {
    const user = await prisma.user.create({
      data: {
        username,
        email,
        hash,
        salt,
      },
    });

    res.status(200).json({
      success: true,
      username: user.username,
      email: user.email,
      message: "Sucess!",
    });
  } catch (err) {
    if (err instanceof Prisma.PrismaClientKnownRequestError) {
      if (err.code === "P2002") {
        return res.status(400).json({
          success: false,
          errors: ["This e-mail is already taken."],
        });
      }
    } else {
      console.log(err);
      return res
        .status(500)
        .json({ success: false, errors: ["Internal Server Error"] });
    }
  }
}
```

To summarize;

(i) I create random users (I check and no two e-mails are the same),

(ii) I send these users to the register API and,

(iii) although no two e-mails are the same, at least 2-3 users do not register and the error "This e-mail is already taken" is returned.

Whatever I did, I couldn't solve the problem. I would be very grateful if you could help me. Thanks in advance.

I even increased the timeout time of requests to the API to 5 seconds, but I still got the same error. I tried many ways but I couldn't solve it.

**Comment by ludralph** [comment]
Hi @periakteon 👋 

Can you please share the repository so I can try reproducing this and debug it?

**Reply by nurul3101** [reply]
@periakteon Please let us know in case you still need any assistance.

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏
